### ISSUE TITLE
Feature Request: waterfall to return promise when callback not defined

### ISSUE THREAD

**Issue by libinvarghese** [caolan/async#1777]
[Currently](https://caolan.github.io/async/v3/docs.html#waterfall), the `waterfall` method expects a callback for the result.

It would be great if `waterfall` method returns a promise when the callback is not provided, like other methods (eg: [`parallel`](https://caolan.github.io/async/v3/docs.html#parallel), [`series`](https://caolan.github.io/async/v3/docs.html#series))

Workaround:

```ts
import { promisify } from "util";
import { waterfall, constant } from "async";

const waterfallP = promisify(waterfall);

const add3 = async (arg1: number): Promise<number> => {
  return Promise.resolve(arg1 + 3);
};

const mul4 = async (arg1: number): Promise<number> => {
  return Promise.resolve(arg1 * 4);
};

const result = await waterfallP([
  constant(1),
  add3,
  mul4
]);

// result = 16
```

**Comment by aearly** [comment]
This could be added, but I wonder why you'd still want to use `waterfall` when you have `async`/`await`.

**Comment by libinvarghese** [comment]
I generally use [compose](https://caolan.github.io/async/v3/docs.html#compose) or [seq](https://caolan.github.io/async/v3/docs.html#seq), to pipe arguments from one async function to the next. But [waterfall](https://caolan.github.io/async/v3/docs.html#waterfall) is more readable.

**Comment by aearly** [comment]
Feel free to open a PR.

**Comment by hargasinski** [comment]
@libinvarghese what version of async are you using? It seems that this is already implemented in the latest version.

**Comment by hargasinski** [comment]
Closed by #1789.
